### PR TITLE
Role switch test

### PR DESCRIPTION
**Pull request by TheTallJerry**
## Motivation and Context

- Role switching returns status 200 regardless of whether the switch was successful. 
- Role switching's error messages could use some improvement. 

## Your Changes

**Description**: 
- Modified role switching so that request status is 404 when trying to role switch to an invalid user, and 422 when trying to role switch to either the current instructor or another instructor. 
- Modified role switching so that the error messages correspond to the correct use case. 
- Added better error message for when trying to role switch to instructors. 

**Type of change** (select all that apply):

- [x] Refactoring (internal change to codebase, without changing functionality)
- [x] Test update (change that modifies or updates tests only)

## Testing

- Written tests covering the entire role switching process. 
- Covered switching as invalid user, current instructor, another instructor, a student/TA. 
- Tested for the session hash and other applicable changes. 

## Questions and Comments (if applicable)

Please help verify that the test breakdown and the added error messages make sense. 

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the pre-commit.ci checks have passed. 
- [x] I have verified that the CI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls.

### CODE REVIEW

**Comment by TheTallJerry** [comment]
I've run i18n-tasks normalize in my container before pushing - what else should I have done so that this i18n test wouldn't have failed?

File: app/controllers/courses_controller.rb [hunk 1]
```diff
@@ -43,27 +43,35 @@ def switch_role
...
-    if found_role.nil? || found_role.instructor?
+    if found_role.nil?
       render partial: 'role_switch_handler',
              formats: [:js], handlers: [:erb],
-             locals: { error: Settings.validate_user_not_allowed_message || I18n.t('main.login_failed') }
+             locals: { error: Settings.validate_user_not_allowed_message || I18n.t('main.login_failed') },
+             status: :not_found
       return
     end
-
```
Comment (david-yz-liu) [inline app/controllers/courses_controller.rb:59]: Revert this change. In general, do not make whitespace changes.

File: app/controllers/courses_controller.rb [hunk 2]
```diff
@@ -43,27 +43,35 @@ def switch_role
...
-    # Check if an instructor trying to login as the current user or themselves
+    # check if the current instructor is trying to role switch as themselves
```
Comment (david-yz-liu) [inline app/controllers/courses_controller.rb:61]: Capitalize comments

File: app/controllers/courses_controller.rb [hunk 3]
```diff
@@ -43,27 +43,35 @@ def switch_role
...
     if found_user.user_name == session[:user_name] || found_user.user_name == session[:real_user_name]
```
Comment (david-yz-liu) [inline app/controllers/courses_controller.rb:62]: Because the `role_switch` policy checks that the user is currently not role-switched, the check against `session[:user_name]` can be removed.

File: app/controllers/courses_controller.rb [hunk 4]
```diff
@@ -43,27 +43,35 @@ def switch_role
...
-      # error
       render partial: 'role_switch_handler',
              formats: [:js], handlers: [:erb],
-             # TODO: put better error message
-             locals: { error: I18n.t('main.login_failed') }
+             locals: { error: I18n.t('main.cannot_role_switch_to_self') },
+             status: :unprocessable_entity
+      return
+    end
+    # otherwise, check if the current instructor is trying to role switch as other instructors
```
Comment (david-yz-liu) [inline app/controllers/courses_controller.rb:69]: Insert a blank line here

File: spec/controllers/courses_controller_spec.rb [hunk 5]
```diff
@@ -1,6 +1,85 @@
 describe CoursesController do
   let(:instructor) { create :instructor }
   let(:course) { instructor.course }
+
+  describe 'role switching methods' do
+    let(:end_user) { create :end_user }
```
Comment (david-yz-liu) [inline spec/controllers/courses_controller_spec.rb:6]: You don't need to define this here. Instead, use a `let` within different context blocks below to define it as an instructor/TA/student. For example, something like

```ruby
context 'when switching to the current instructor' do
  let(:end_user) { instructor }
  it ...
```

Specifically, the `let` statement for `subject` can appear *before* the `let` statement for `end_user`, even if `subject` is defined using `end_user`.

File: spec/controllers/courses_controller_spec.rb [hunk 6]
```diff
@@ -1,6 +1,85 @@
...
+    let(:subject) do
+      post_as instructor, :switch_role, params: { 'id' => course.id, 'effective_user_login' => end_user.user_name }
```
Comment (david-yz-liu) [inline spec/controllers/courses_controller_spec.rb:8]: In general use symbol keys, not strings. So `id: course.id` instead of `'id' => course.id`, for example.

File: spec/controllers/courses_controller_spec.rb [hunk 7]
```diff
@@ -1,6 +1,85 @@
...
+    end
+    describe '#switch_role' do
+      let(:second_instructor) { create :instructor }
```
Comment (david-yz-liu) [inline spec/controllers/courses_controller_spec.rb:11]: Set the course association for instructors/TAs/students you create here, to be explicit that they are part of the same course.

(Unless you are testing that they aren't. In fact, you should add tests checking that `:forbidden` status is returned if trying to switch role to a role in a different course.

File: spec/controllers/courses_controller_spec.rb [hunk 8]
```diff
@@ -1,6 +1,85 @@
...
+      let(:temp_user) { create :end_user }
+
+      it 'fails when no username is provided ' do
+        post_as instructor, :switch_role, params: { 'id' => course.id }
+        expect(response).to have_http_status(:not_found)
+      end
+      it 'fails when an invalid username is provided' do
+        User.delete(temp_user)
```
Comment (david-yz-liu) [inline spec/controllers/courses_controller_spec.rb:19]: You can do `temp_user.destroy` here.

File: spec/controllers/courses_controller_spec.rb [hunk 9]
```diff
@@ -1,6 +1,85 @@
...
+        post_as instructor, :switch_role, params: { 'id' => course.id, 'effective_user_login' => temp_user.user_name }
+        expect(response).to have_http_status(:not_found)
+      end
+      it 'fails when switching role as the current instructor' do
+        end_user.roles << instructor
```
Comment (david-yz-liu) [inline spec/controllers/courses_controller_spec.rb:24]: Remove this line; see above comment about using `let` to define `end_user`

File: spec/controllers/courses_controller_spec.rb [hunk 10]
```diff
@@ -1,6 +1,85 @@
...
+        expect(subject).to have_http_status(:unprocessable_entity)
+      end
+      it 'fails when switching role as another instructor in the course' do
+        end_user.roles << second_instructor
+        expect(subject).to have_http_status(:unprocessable_entity)
+      end
+      context 'when switching to a student' do
```
Comment (david-yz-liu) [inline spec/controllers/courses_controller_spec.rb:31]: Also check for the response status and `session[:user_name]` here (and for the TA tests below)

File: spec/controllers/courses_controller_spec.rb (hunk 7 above)
Comment (TheTallJerry) [reply spec/controllers/courses_controller_spec.rb:11]: We currently do not have an option of returning the :forbidden or 403 status. Right now the routine is: 
- First we accquire the username. If it's blank we say username can't be blank and return a 404. 
- Then we try to find the user in the db + find their role in the course. If their role is nil we say login failed and return a 404
- Then we check whether they are logging in as themselves or another instructor in the course. Either case will be 422 but we'll give different messages
- Finally, we return 200 and do the stuff afterwards. 
I'm adding the tests for if they are switching to valid users not in the course - but let me know how you would like the status to be handled (I can either keep using 404 as their role will be nil as well, or I can modify the controller to use 403 with an extra if block).

Comment (david-yz-liu) [reply spec/controllers/courses_controller_spec.rb]: @TheTallJerry ah yes I see, a `:not_found` is totally fine. Thanks for the explanation. You can finalize the tests to check for `:not_found` (I guess just remove the parts you commented out.)

**Review by david-yz-liu — APPROVED** [review]
